presto-main/src/main/java/com/facebook/presto/sql/analyzer/MaterializedViewColumnMappingExtractor.java

```diff
@@ -0,0 +1,131 @@
+import static com.facebook.presto.spi.ConnectorMaterializedViewDefinition.TableColumn;
+import static com.facebook.presto.sql.analyzer.SemanticErrorCode.NOT_SUPPORTED;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.util.Objects.requireNonNull;
+
+public class MaterializedViewColumnMappingExtractor
+        extends MaterializedViewPlanValidator
+{
+    private final Analysis analysis;
+
+    private ImmutableList.Builder<List<TableColumn>> mappedBaseColumns;
+
+    private MaterializedViewColumnMappingExtractor(Analysis analysis)
+    {
+        this.analysis = requireNonNull(analysis, "analysis is null");
+
+        this.mappedBaseColumns = ImmutableList.builder();
+    }
+
+    public static List<List<TableColumn>> extractMappedBaseColumns(Analysis analysis)
+    {
+        checkState(
+                analysis.getStatement() instanceof CreateMaterializedView,
+                "Only support extracting of mapped columns from create materialized view query");
+
+        MaterializedViewColumnMappingExtractor extractor = new MaterializedViewColumnMappingExtractor(analysis);
+        Query viewQuery = ((CreateMaterializedView) analysis.getStatement()).getQuery();
+        extractor.process(viewQuery, new MaterializedViewPlanValidatorContext());
+
+        return extractor.getMappedBaseColumns();
+    }
+
+    @Override
+    protected Void visitSetOperation(SetOperation node, MaterializedViewPlanValidatorContext context)
+    {
+        super.visitSetOperation(node, context);
+
+        List<RelationType> outputDescriptorList = node.getRelations().stream()
+                .map(relation -> analysis.getOutputDescriptor(relation).withOnlyVisibleFields())
+                .collect(toImmutableList());
+
+        int numRelations = outputDescriptorList.size();
+        int numFields = outputDescriptorList.get(0).getVisibleFieldCount();
+        for (int fieldIndex = 0; fieldIndex < numFields; fieldIndex++) {
+            for (int firstRelationIndex = 0; firstRelationIndex < numRelations; firstRelationIndex++) {
+                Optional<TableColumn> firstBaseColumn = tryGetOriginalTableColumn(outputDescriptorList.get(firstRelationIndex).getFieldByIndex(fieldIndex));
+                for (int secondRelationIndex = firstRelationIndex + 1; secondRelationIndex < numRelations; secondRelationIndex++) {
+                    Optional<TableColumn> secondBaseColumn = tryGetOriginalTableColumn(outputDescriptorList.get(secondRelationIndex).getFieldByIndex(fieldIndex));
+                    if (firstBaseColumn.isPresent() && secondBaseColumn.isPresent() && !firstBaseColumn.get().equals(secondBaseColumn.get())) {
+                        mappedBaseColumns.add(ImmutableList.of(firstBaseColumn.get(), secondBaseColumn.get()));
+                    }
+                }
+            }
+        }
```

Comment:
This solution has very high time complexity and space complexity.
The mappedBaseColumns is used to find all the mapped based columns for a given column. This is a typical union-find problem. A union-find solution would be much more efficient and cheaper.

Reply:
This is just an optimization and doesn't affect the functionality. Considering that it requires significant changes to switch to a union-find solution, I am fine with adding a TODO here and optimize it in a following PR.

Reply:
I actually can see a trade-off between time and space here; that is, instead of returning the mapped base columns of a given MV column as a list of enumerated pairs, we could simply return lists of base columns. Theoretically it is a union-find or CC problem. But I think we might not need to go that far, since in practice there might not be not much overlap among the lists.